File: docker-compose/vault/config/unseal.py
#!/usr/bin/env python

import json
import subprocess
import os
import signal

'''
This script will unseal vault server and get a periodic token.

manually run this script when...
1. vault server restart.
2. vault server start.

After get a periodic token, copy and paste to server env file.
'''

base_path = "/vault"
file_path = os.path.join(base_path, "file")
config_path = os.path.join(base_path, "config")
policies_path = os.path.join(base_path, "policies")

db_policy_file = os.path.join(policies_path, "db_hcl.sh")
db_connection_file = os.path.join(config_path, "connection_db.sh")
rule_sql_file = os.path.join(config_path, "create_user.sql")

db_name = os.getenv("POSTGRES_BACKEND_DB")
cred_name = os.getenv("VAULT_CRED_NAME")
policy_name = os.getenv("VAULT_POLICY_NAME")

def main():
    pipe = subprocess.Popen('vault operator init -format=json -status=true', shell=True, stdout=subprocess.PIPE)
    unseal_status = pipe.communicate()[0]
    unseal_status = json.loads(unseal_status)
    unseal_status = unseal_status['Initialized']

    unseal_keys = []
    root_token = None

    if unseal_status == False:
        pipe = subprocess.Popen(f'vault operator init -format=json', shell=True, stdout=subprocess.PIPE)
        init_results = pipe.communicate()[0]
        init_results = json.loads(init_results)
        unseal_keys = init_results['unseal_keys_b64']
        root_token = init_results['root_token']
        print('================================unseal_keys========================================')
        print(unseal_keys)
        print('================================unseal_keys========================================\n\n')
        print('================================root_token========================================')
        print(root_token)        
        print('================================root_token========================================\n\n')
    else:
        init_results = input("Please input unseal keys in a list. : \n")
        init_results = json.loads(init_results.replace("'", '"'))
        unseal_keys = init_results
        root_token = input("Please input root token. : \n")

    for key, unseal_key in enumerate(unseal_keys):
        if key >= 3:
            break
        subprocess.call(f'vault operator unseal {unseal_key}', shell=True)

    subprocess.call(f'vault login {root_token}', shell=True)
    subprocess.call('vault secrets enable database', shell=True)

    subprocess.call(f'{db_policy_file} | vault policy write {policy_name} -', shell=True)
    subprocess.call(f'{db_connection_file} | vault write database/config/{db_name} -', shell=True)
    # subprocess.call(f'vault write -force database/rotate-root/{db_name}', shell=True)
    subprocess.call(f'vault write database/roles/{cred_name} db_name={db_name} creation_statements=@{rule_sql_file} default_ttl=1h max_ttl=24h', shell=True)
    
    pipe = subprocess.Popen(f'vault token create -policy={policy_name} -period 1h -format=json', shell=True, stdout=subprocess.PIPE)
    db_token_results = pipe.communicate()[0]
    db_token_results = json.loads(db_token_results)

    print('======================db_token==============================')
    print(db_token_results['auth']['client_token'])
    print('======================db_token==============================')

if __name__ == '__main__':
    try:
        main()
    except Exception:
        # no matter what error happen....just kill process 1 (which means, kill the vault server.)
        os.kill(1, signal.SIGTERM)

File: docker-compose/vault/config/unseal.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"syscall"
)

// This script will unseal vault server and get a periodic token.

// manually run this script when...
// 1. vault server restart.
// 2. vault server start.

// After get a periodic token, copy and paste to server env file.

func unseal() {
	const basePath = "/vault"
	configPath := filepath.Join(basePath, "config")
	policiesPath := filepath.Join(basePath, "policies")

	dbPolicyFile := filepath.Join(policiesPath, "db_hcl.sh")
	dbConnectionFile := filepath.Join(configPath, "connection_db.sh")
	ruleSqlFile := filepath.Join(configPath, "create_user.sql")

	dbName := os.Getenv("POSTGRES_BACKEND_DB")
	credName := os.Getenv("VAULT_CRED_NAME")
	policyName := os.Getenv("VAULT_POLICY_NAME")

	cmd := "vault operator init -format=json -status=true"
	out, _ := exec.Command("sh", "-c", cmd).Output()
	var unsealStatus map[string]interface{}
	json.Unmarshal(out, &unsealStatus)

	var unsealKeys []interface{}
	var rootToken string

	if unsealStatus["Initialized"].(bool) == false {
		cmd = "vault operator init -format=json"
		out, _ = exec.Command("sh", "-c", cmd).Output()
		var initResults map[string]interface{}
		json.Unmarshal(out, &initResults)
		unsealKeys = initResults["unseal_keys_b64"].([]interface{})
		rootToken = initResults["root_token"].(string)
		fmt.Println("=========================Unseal Keys====================================")
		fmt.Println(unsealKeys)
		fmt.Println()
		
		fmt.Println("=========================Root Token====================================")
		fmt.Println(rootToken)
		fmt.Println()
		
	} else {
		fmt.Println("Please enter unseal keys in a slice.")
		var input string
    	fmt.Scanln(&input)
		var initResults []interface{}
		json.Unmarshal([]byte(input), &initResults)
		unsealKeys = initResults
		
		fmt.Println("Please enter root token.")
		fmt.Scanln(&rootToken)
	}

	for index, unsealKey := range unsealKeys {
		if index >= 3 {
			break
		}
		cmd = fmt.Sprintf("vault operator unseal %s", unsealKey)
		_, err := exec.Command("sh", "-c", cmd).Output()
		if err != nil {
			panic("Fail to unseal vault server.")
		}
	}

	cmd = fmt.Sprintf("vault login %s", rootToken)
	_, err := exec.Command("sh", "-c", cmd).Output()
	if err != nil {
		panic("Fail to login with root token.")
	}

	cmd = "vault secrets enable database"
	exec.Command("sh", "-c", cmd).Output()

	cmd = fmt.Sprintf("%s | vault policy write %s -", dbPolicyFile, policyName)
	exec.Command("sh", "-c", cmd).Output()

	cmd = fmt.Sprintf("%s | vault write database/config/%s -", dbConnectionFile, dbName)
	exec.Command("sh", "-c", cmd).Output()

	// cmd = fmt.Sprintf("vault write -force database/rotate-root/%s", dbName)
	// exec.Command("sh", "-c", cmd).Output()

	cmd = fmt.Sprintf("vault write database/roles/%s db_name=%s creation_statements=@%s default_ttl=1h max_ttl=24h", credName, dbName, ruleSqlFile)
	exec.Command("sh", "-c", cmd).Output()

	cmd = fmt.Sprintf("vault token create -policy=%s -period 1h -format=json", policyName)
	out, _ = exec.Command("sh", "-c", cmd).Output()
	var dbTokenResults map[string]interface{}
	json.Unmarshal(out, &dbTokenResults)

	fmt.Println("\n=========================DB Token====================================")
	fmt.Println(dbTokenResults["auth"].(map[string]interface{})["client_token"])
	fmt.Println()
}

func doUnseal(fn func()) {
	defer func() {
		if err := recover(); err != nil {
			fmt.Println(err)
			syscall.Kill(1, syscall.SIGTERM)
		}
	}()
	fn()
}

func main() {
	doUnseal(unseal)
}